增加feature,修改bug

wulaxx opened the pull request:
new feature:
TcpClient&HttpClient中新增两个接口:Connect(const char _url ...)和Connect(const base::Uri& url ...),通过getaddrinfo获取解析的IP,对于有多个IP的情况逐个尝试,知道有一个连接成功,如果都连接失败返回-1;
bug fix:
EndPoint::FromSockAddr(sockaddr_ addr, socklen_t len)中检查len使用的枚举常量不对,替换成sizeof(sockaddr_in)和sizeof(sockaddr_in6);
examples:
修改了client.cc的测试代码,增加了命令行参数url;

iElaine committed:
TcpClient和HttpClient添加了两个Connect方法,支持dns;修改了EndPoint中FromSockAddr的bug

iElaine committed:
添加EndPoint的部分单元测试代码

iElaine committed:
去掉client.cc中打印的调试信息

myjfm commented on src/tcp/tcp_client.cc:
```diff
@@ -109,6 +111,52 @@ ConnectionId TcpClient::Connect(const base::EndPoint* remote,
   return connection->id();
 }
 
+ConnectionId TcpClient::Connect(const char *url,
```
TcpClient不需要感知url，只需要知道ip和port就可以了

myjfm commented on src/http/http_client.h:
```diff
@@ -49,6 +50,12 @@ class HttpClient final : public HttpBase {
+  tcp::ConnectionId Connect(const char *url,
+                       const HttpOptions& options);
+
+  tcp::ConnectionId Connect(const base::Uri& url,
```
HttpClient只需要加一个接口：
tcp::ConnectionId Connect(base::StringPiece url, const HttpOptions& options);接口就行了
不需要const char*, 也不需要base::Uri

iElaine committed:
1.删除了TcpClient中通过域名连接的借口,HttpClient
接口改为Connect(base::StringPiece url, ...);2. Uri中如果不指定端口,会根据Scheme的默认端口制定port_,增加了port_str_以保证解析前后url的一致性

wulaxx commented:
1. 删除了TcpClient中通过域名连接的接口,HttpClient接口改为Connect(base::StringPiece url, ...);
2. Uri中如果不指定端口,会根据Scheme的默认端口制定port_,增加了port_str_以保证解析前后url的一致性;

iElaine committed:
增加Uri的部分单元测试代码

iElaine committed:
修改了uri中的get_scheme_default_port的bug

iElaine committed:
调整Uri::Parse中解析port的部分代码

wulaxx commented:
改好了

iElaine committed:
修改TcpClient和HttpClient失败的返回值为kInvalidConnectionId

myjfm commented on src/base/uri.cc:
```diff
@@ -49,6 +50,15 @@ const std::regex Uri::kAuthorityRegex(
 const std::regex Uri::kQueryParamRegex("(^|&)([^=&]*)=?([^=&]*)(?=(&|$))");
 
+static const std::unordered_map<std::string, uint16_t> kSchemeToPortMap = {{"http", 80}, {"https", 443}};
+static uint16_t get_scheme_default_port(const std::string& schema) {
```
get_scheme_defualt_port函数每个单词首字母大写，不用_连接

myjfm commented on src/http/http_client.cc:
```diff
@@ -61,6 +62,35 @@ tcp::ConnectionId HttpClient::Connect(const base::EndPoint* remote,
+  struct addrinfo *presults = nullptr;
+  struct addrinfo hint;
+  bzero(&hint, sizeof(hint));
+  hint.ai_family = AF_INET;
```
不需要填AF_INET，保持默认值就行，我们可以支持ipv6

myjfm replied:
至少将来是需要支持ipv6的

myjfm commented on src/http/http_client.cc:
```diff
@@ -61,6 +62,35 @@ tcp::ConnectionId HttpClient::Connect(const base::EndPoint* remote,
+    return tcp::kInvalidConnectionId;
+  }
+  base::EndPoint endpoint;
+  if (!endpoint.FromSockAddr(*presults->ai_addr, presults->ai_addrlen)) {
```
return之前还是需要freeaddrinfo的吧

myjfm commented on src/http/http_client.h:
```diff
@@ -49,6 +50,8 @@ class HttpClient final : public HttpBase {
   tcp::ConnectionId Connect(const base::EndPoint* remote,
                             const HttpOptions& options);
+  tcp::ConnectionId Connect(base::StringPiece url,
+                       const HttpOptions& options);
```
缩进保持统一风格，和第一个参数对齐

myjfm commented on unittests/end_point_unittest.cc:
```diff
@@ -0,0 +1,13 @@
+#include <base/end_point.h>
+#include <netinet/in.h>
+TEST(EndPoint, SockAddrTest)
+{
```
{和TEST放到同一行

myjfm commented on unittests/uri_unittest.cc:
```diff
@@ -0,0 +1,22 @@
+#include <base/uri.h>
+
+TEST(Uri, ParseTest)
+{
```
{和TEST放到同一行

iElaine committed:
修改HttpClient的Connect,使之解析的域名地址也包含IPV6的

iElaine committed:
Merge branch 'hotfix/name_resolve_support_ipv6'

修改了HttpClient::Connect的实现,使解析的域名地址中也包含ipv6的地址

iElaine committed:
统一风格,修正Connect失败没有释放addrinfo的bug

修改超过80行的部分

iElaine committed:
统一风格,修正Connect失败没有释放addrinfo的bug

iElaine committed:
Merge branch 'hotfix/connect_bug_fix'

myjfm commented on src/base/uri.cc:
```diff
@@ -49,6 +50,15 @@ const std::regex Uri::kAuthorityRegex(
 
 const std::regex Uri::kQueryParamRegex("(^|&)([^=&]*)=?([^=&]*)(?=(&|$))");
 
+static const std::unordered_map<std::string, uint16_t> kSchemeToPortMap = {{"http", 80}, {"https", 443}};
```
一行不超过80个字符

iElaine committed:
Merge branch 'hotfix/connect_bug_fix'

Conflicts:
	src/base/uri.cc
